user_profile/views.py

```diff
@@ -0,0 +1,143 @@
+
+from .models import Profile, Skill, Project
+from .forms import ProfileForm, ProjectForm
+
+
+class DetailSkillView(DetailView):
+    model = Skill
+
+
+class DetailProjectView(DetailView):
+    model = Project
+
+
+class ListProjectView(ListView):
+    model = Project
+
+
+class CreateProjectView(PermissionRequiredMixin, CreateView):
+    redirect_field_name = '/'
+    permission_required = 'user_profile.add_project'
+    model = Project
+    form_class = ProjectForm
+
+
+class EditProjectView(UserPassesTestMixin, UpdateView):
+    model = Project
+    form_class = ProjectForm
+    redirect_field_name = '/'
+
+    def test_func(self):
+        return self.request.user == get_object_or_404(Project, pk=self.kwargs['pk']).manager
+
+
+class DeleteProjectView(UserPassesTestMixin, DeleteView):
+    model = Project
+    redirect_field_name = '/'
+    success_url = reverse_lazy('project_list')
+
+    # Should be same the same in DeleteProject and EditProject
+    def test_func(self):
+        return self.request.user == get_object_or_404(Project, pk=self.kwargs['pk']).manager
+
+##############################################
+
+
+def administrate_project(request, pk):
+    project = get_object_or_404(Project, pk=pk)
+    if not request.user == project.manager:
+        return HttpResponseRedirect("/")
+
+    return render(request, 'user_profile/project_admin.html', {'project': project})
+
+
+def accept_applicant(request, pk, username):
+    return manage_applicant(request, pk, username, True)
+
+
+def reject_applicant(request, pk, username):
+    return manage_applicant(request, pk, username, False)
+
+
+def apply_to_project(request, pk):
+    if request.method == 'POST':
+        user = get_object_or_404(User, username=request.user.username)
+        project = get_object_or_404(Project, pk=pk)
+
+        # if not rejected, already an applicant or already a member
+        if user in project.applicants.all() or \
+                user in project.members.all() or\
+                user in project.rejected_applicants.all():
+            return HttpResponse("You have already applied to this project")
+        try:
+            if project.application_end < datetime.date.today():
+                return HttpResponse("Applications have ended for this project")
+        except TypeError:
+            return HttpResponse("This project does not have an application date set")
+
+        project.applicants.add(user)
+        project.save()
+        messages.success(request, 'You have successfully applied to ' + str(project))
+        return HttpResponseRedirect(reverse('project', kwargs={'pk': pk}))
+
+    return HttpResponseRedirect('/')
+
+
+def profile(request, username):
+    user = get_object_or_404(User, username=username)
+    try:
+        user.profile  # Accessing a non-existent profile (they do no exist by default) triggers an error
+    except Profile.DoesNotExist:
```

Comment:
Profile.objects.create(user=user)

Reply:
Done. Doesn't show here for some reason.